src/com/teamcautionrobotics/robot2018/Lift.java

```diff
@@ -0,0 +1,161 @@
+            } else {
+                return values[this.ordinal() - 1];
+            }
+        }
+    }
+
+    private VictorSP liftMotor;
+    private Encoder liftEncoder;
+    private PIDController pidController;
+
+    private double destinationHeight;
+
+    public Lift(int motorPort, int encoderChannelA, int encoderChannelB, double Kp, double Ki,
+            double Kd) {
+        liftMotor = new VictorSP(motorPort);
+        liftEncoder = new Encoder(encoderChannelA, encoderChannelB);
+        liftEncoder.setDistancePerPulse((4 * Math.PI) / 1024);
+        pidController = new PIDController(Kp, Ki, Kd, 0, this, this);
+        pidController.setOutputRange(-1, 1);
+        pidController.setAbsoluteTolerance(3);
+    }
+
+    /**
+     * @param power positive is ascending, negative is descending, range of [-1, 1]
+     */
+    public void move(double power) {
+        double rateOfChange = power * 10; //inches per second
+        double dt = 0.02; //seconds
+        pidController.setSetpoint(getCurrentHeight() + rateOfChange * dt);
+    }
+
+    public void ascend() {
+        this.move(1.0);
+    }
+
+    public void descend() {
+        this.move(-1.0);
+    }
+
+    public void stop() {
+        this.move(0);
+    }
+
+    public void setLevel(LiftLevel destinationLiftLevel) {
+        destinationHeight = destinationLiftLevel.height;
+        pidController.setSetpoint(destinationHeight);
+        enablePID();
+    }
+
+    public void setHeight(double height) {
+        pidController.setSetpoint(height);
+        enablePID();
+    }
+
+    public LiftLevel getCurrentLiftLevel() {
+        return convertHeightToLiftLevel(getCurrentHeight());
+    }
+
+    public double getCurrentHeight() {
+        return getDistance();
+    }
+
+    public double getDestinationHeight() {
+        return pidController.getSetpoint();
+    }
+
+    public LiftLevel getDestinationLiftLevel() {
+        return convertHeightToLiftLevel(getDestinationHeight());
+    }
+
+    private LiftLevel convertHeightToLiftLevel(double height) {
+        LiftLevel convertedLiftLevel = LiftLevel.GROUND;
+        for (LiftLevel liftLevel : LiftLevel.values) {
+            double midpoint = (liftLevel.height + liftLevel.next().height) / 2;
+            convertedLiftLevel = liftLevel;
+            if (height < midpoint) {
+                break;
+            }
+        }
+        return convertedLiftLevel;
+    }
+
+    public void enablePID() {
+        if (!pidController.isEnabled()) {
+            pidController.enable();
+        }
+    }
+
+    public void disablePID() {
+        if (pidController.isEnabled()) {
+            pidController.disable();
+        }
+    }
+
+    public void resetEncoder() {
+        liftEncoder.reset();
+    }
+
+    public double getDistance() {
+        return liftEncoder.getDistance();
+    }
+
+    @Override
+    public void pidWrite(double height) {
+        SmartDashboard.putNumber("pid lift height", height);
+    }
+
+    @Override
```

Comment:
Instead of Lift implementing PIDSource, have a subclass of Lift that implements PIDSource. This subclass should also support `setPIDSourceType(PIDSourceType)`.